Plugin/Framework/Search/Request/Builder.php

```diff
@@ -0,0 +1,83 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2020 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Cmp\Plugin\Framework\Search\Request;
+
+use Magento\Framework\ObjectManagerInterface;
+use Magento\Framework\Search\Request\Builder as MagentoRequestBuilder;
+use Magento\Framework\Search\Request\Query\BoolExpression;
+use Magento\Framework\Search\Request\QueryInterface;
+use Magento\Framework\Search\RequestInterface;
+use Nosto\Cmp\Model\Search\Request as NostoSearchRequest;
+use Nosto\Cmp\Utils\Request as RequestUtils;
+
+class Builder
+{
+    /**
+     * @var ObjectManagerInterface
+     */
+    private $objectManager;
+
+    /**
+     * Builder constructor.
+     * @param ObjectManagerInterface $objectManager
+     */
+    public function __construct(ObjectManagerInterface  $objectManager)
+    {
+        $this->objectManager = $objectManager;
+    }
+
+    /**
+     * @param MagentoRequestBuilder $builder
+     * @param RequestInterface $request
+     * @return RequestInterface
+     * @noinspection RedundantSuppression
+     */
+    // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter
+    public function afterCreate(MagentoRequestBuilder $builder, RequestInterface $request)
+    {
+        $query = $request->getQuery();
+        if ($query instanceof BoolExpression && RequestUtils::containsBoolNostoSearchQuery($query)) {
+            /** @var NostoSearchRequest $nostoRequest */
+            $nostoRequest =  $this->objectManager->create(
```

Comment:
Why can't you simply wrap **all** queries - why does it need to be a bool query or a Nosto query? Would it really maatter?